qualtran/bloqs/arithmetic/permutation.py

```diff
@@ -0,0 +1,248 @@
+    if isinstance(cycle, Shaped):
+        return cycle
+    return tuple(cycle)
+
+
+@frozen
+class PermutationCycle(Bloq):
+    r"""Apply a single permutation cycle on the basis states.
+
+    Args:
+        N: the total size the permutation acts on.
+        cycle: the permutation cycle to apply.
+
+    Registers:
+        x: integer register storing a value in [0, ..., N - 1]
+
+    References:
+        [A simple quantum algorithm to efficiently prepare sparse states](https://arxiv.org/abs/2310.19309v1)
+        Appendix B, Algorithm 7.
+    """
+
+    N: SymbolicInt
+    cycle: Union[tuple[int, ...], Shaped] = field(converter=_convert_cycle)
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature.build_from_dtypes(x=BoundedQUInt(self.bitsize, self.N))
+
+    @cached_property
+    def bitsize(self):
+        return bit_length(self.N - 1)
+
+    def is_symbolic(self):
+        return is_symbolic(self.N, self.cycle)
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', x: 'SoquetT') -> dict[str, 'SoquetT']:
+        if self.is_symbolic():
+            raise DecomposeTypeError(f"cannot decompose symbolic {self}")
+        assert not isinstance(self.cycle, Shaped)
+
+        a: 'SoquetT' = bb.allocate(dtype=QBit())
+
+        for k, x_k in enumerate(self.cycle):
+            x, a = bb.add_t(EqualsAConstant(self.bitsize, x_k), x=x, target=a)
+
+            delta = x_k ^ self.cycle[(k + 1) % len(self.cycle)]
+            a, x = bb.add_t(XorK(QUInt(self.bitsize), delta).controlled(), ctrl=a, x=x)
+
+        x, a = bb.add_t(EqualsAConstant(self.bitsize, self.cycle[0]), x=x, target=a)
+
+        bb.free(cast(Soquet, a))
+
+        return {'x': x}
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        if self.is_symbolic():
+            x = ssa.new_symbol('x')
+            cycle_len = slen(self.cycle)
+            return {
+                (EqualsAConstant(self.bitsize, x), cycle_len + 1),
+                (XorK(QUInt(self.bitsize), x).controlled(), cycle_len),
+            }
+
+        return super().build_call_graph(ssa)
+
+
+@bloq_example
+def _permutation_cycle() -> PermutationCycle:
+    permutation_cycle = PermutationCycle(4, (0, 1, 2))
+    return permutation_cycle
+
+
+@bloq_example
+def _permutation_cycle_symb() -> PermutationCycle:
+    import sympy
+
+    N, L = sympy.symbols("N L")
+    cycle = Shaped((L,))
+    permutation_cycle_symb = PermutationCycle(N, cycle)
+    return permutation_cycle_symb
+
+
+def _convert_cycles(cycles) -> Union[tuple[SymbolicCycleT, ...], Shaped]:
+    if isinstance(cycles, Shaped):
+        return cycles
+    return tuple(_convert_cycle(cycle) for cycle in cycles)
+
+
+@frozen
+class Permutation(Bloq):
+    """Apply a permutation of [0, N - 1] on the basis states.
+
+    Decomposes a permutation into cycles and applies them in order.
+    See :meth:`from_dense_permutation` to construct this bloq from a permutation.
```

Comment:
Ditto: Please add a latex equation describing the action of the unitary on an input state.